### PR TITLE
Validate and update manage applications

### PR DESCRIPTION
**Pull request by Zain-Abbas1**
- match update app page to prototype
- add working service to interact with api
- update validator
- add all relevant unit tests

- FA-786

### CODE REVIEW

File: server/routes/applications.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ router.get('(/manage-applications)?', (req, res, next) => {
       test: listTest.data,
       future: listFuture.data
     };
+    console.log('GET LIST ALL: ', viewData.this_data.test);
```
Comment (Zain-Abbas1) [inline server/routes/applications.js:33]: I will remove this

File: server/routes/applications.js [hunk 2]
```diff
@@ -96,15 +97,55 @@ router.get('/manage-applications/:appId/view/:env', (req, res, next) => {
...
+    title: 'Edit application'
+  };
+
+  applicationsDeveloperService.getApplication(id, env)
+    .then(appData => {
+      viewData.this_data.applicationName = appData.data.name;
+      viewData.this_data.description = appData.data.description;
+      viewData.this_data.privacyPolicy = appData.data.privacy_policy_url;
+      viewData.this_data.terms = appData.data.terms_and_conditions_url;
+      console.log('APP DATA: ', appData);
```
Comment (Zain-Abbas1) [inline server/routes/applications.js:120]: Will remove

File: server/routes/applications.js [hunk 3]
```diff
@@ -96,15 +97,55 @@ router.get('/manage-applications/:appId/view/:env', (req, res, next) => {
...
   };
-  res.render(`${routeViews}/edit.njk`, viewData);
+  validator.validateApplication(payload, true)
+    .then(_ => {
+      return applicationsDeveloperService.updateApplication(payload);
+    }).then(_ => {
+      return res.redirect(302, '/manage-applications');
+    }).catch(err => {
+      viewData.this_errors = routeUtils.processException(err);
+      console.log(viewData.this_errors);
```
Comment (Zain-Abbas1) [inline server/routes/applications.js:146]: Will remove

File: server/lib/validation/form_validators/ManageApplication.js [hunk 4]
```diff
@@ -36,8 +36,10 @@ class ManageApplication extends Validator {
           this.errors.description = errorManifest.description.invalid;
         }
       }
-      if (typeof payload.environment === 'undefined' || payload.environment === '') {
-        this.errors.environment = errorManifest.environment.blank;
+      if (isUpdate === false) {
+        if (typeof payload.environment === 'undefined' || payload.environment === '') {
+          this.errors.environment = errorManifest.environment.blank;
+        }
```
Comment (CH-JonathanBlake) [inline server/lib/validation/form_validators/ManageApplication.js:42]: Would it make more sense to set up different validation methods for post and update?

File: server/routes/applications.js (hunk 1 above)
Comment (CH-JonathanBlake) [reply server/routes/applications.js]: You just left that in to test us of course. ;)

File: server/routes/applications.js [hunk 5]
```diff
@@ -96,15 +97,55 @@ router.get('/manage-applications/:appId/view/:env', (req, res, next) => {
...
+    this_data: payload,
     this_errors: null,
     active_page: 'application-overview',
-    title: 'Manage Application'
+    title: 'Update an application'
   };
-  res.render(`${routeViews}/edit.njk`, viewData);
+  validator.validateApplication(payload, true)
+    .then(_ => {
+      return applicationsDeveloperService.updateApplication(payload);
```
Comment (CH-JonathanBlake) [inline server/routes/applications.js:139]: I'd rather see things passed in as specific function parameters. I don't like that we have to assume the structure of the 'payload' in the update method and we don't have anything to ensure that values are passed in as expected.

Comment (CH-JonathanBlake) [reply server/routes/applications.js:139]: So updateApplication( payload, appId, env)
This is more inline with how its been done in other methods of the service class.

**Review by CH-JonathanBlake — CHANGES_REQUESTED** [review]

**Comment by CH-JonathanBlake** [comment]
I don't know if those changes are going to get push back as the team might prefer your way.

File: server/routes/applications.js (hunk 5 above)
Comment (CH-JonathanBlake) [reply server/routes/applications.js:139]: Agreed to disagree. And not refactor other functions to match

File: server/lib/validation/form_validators/ManageApplication.js (hunk 4 above)
Comment (Zain-Abbas1) [reply server/lib/validation/form_validators/ManageApplication.js]: Refactored validate method out to addApplication & updateApplication

File: test/server/lib/validation/form_validators/ManageApplications.test.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ describe('server/lib/validation/form_validators/ManageApplication', () => {
     const stubValidDesc = sinon.stub(Validator.prototype, 'isValidDescription').returns(true);
     const stubValidUrl = sinon.stub(Validator.prototype, 'isValidUrl').returns(true);
     const stubKeys = sinon.stub(ManageApplication.prototype, '_formatIncomingPayload').returns(routeData.addApplication);
-    expect(ManageApplication.prototype.validateApplication(payload, false)).to.eventually.equal(true);
+    expect(ManageApplication.prototype.addApplication(payload)).to.eventually.equal(true);
```
Comment (CH-JonathanBlake) [inline test/server/lib/validation/form_validators/ManageApplications.test.js]: Thinking about this I wonder if we should have moved most of these tests to update and then stubbed that and just kept the env tests on add.

**Review by CH-JonathanBlake — APPROVED** [review]
Changes made as requested. Possible testing question but not a blocker.

File: test/server/lib/validation/form_validators/ManageApplications.test.js (hunk 6 above)
Comment (Zain-Abbas1) [reply test/server/lib/validation/form_validators/ManageApplications.test.js]: We could have, it's a slight adjustment I'm happy to leave as it is.

**Review by mwejuli — APPROVED** [review]

**Pull request merged**